File: windows/src/HoverPocket.Shell/PocketApps/PocketAppRuntimeActivationRegistry.cs
using HoverPocket.Shell.Capabilities;
using HoverPocket.Shell.Configuration;

namespace HoverPocket.Shell.PocketApps;

internal sealed record PocketAppRuntimeReadback(
    string AppId,
    string? Version,
    string? PackageDigest,
    IReadOnlyList<string> EffectivePermissions)
{
    public bool Matches(PocketAppLifecycleReceipt receipt) =>
        string.Equals(AppId, receipt.PackageId, StringComparison.Ordinal)
        && string.Equals(Version, receipt.Version, StringComparison.Ordinal)
        && string.Equals(PackageDigest, receipt.PackageDigest, StringComparison.Ordinal)
        && EffectivePermissions.Order(StringComparer.Ordinal).SequenceEqual(
            receipt.EffectivePermissions.Order(StringComparer.Ordinal),
            StringComparer.Ordinal);
}

internal sealed class PocketAppRuntimeActivationException(string code) : Exception(code)
{
    public string Code { get; } = code;
}

internal sealed class PocketAppActivationLease
{
    private int _active = 1;
    private readonly CancellationTokenSource _cancellation = new();

    public bool IsActive => Volatile.Read(ref _active) == 1;

    public CancellationToken CancellationToken => _cancellation.Token;

    public void RequireActive()
    {
        if (!IsActive)
        {
            throw new PocketAppRuntimeActivationException("RUNTIME_ACTIVATION_UNAVAILABLE");
        }
    }

    public void Invalidate()
    {
        if (Interlocked.Exchange(ref _active, 0) == 1)
        {
            _cancellation.Cancel();
        }
    }
}

internal sealed class PocketExecutionRuntimeRegistry
{
    private sealed record Entry(
        PocketAppRuntimeReadback Readback,
        object RuntimeHandle,
        PocketAppActivationLease? ActivationLease);
    private readonly Dictionary<string, Entry> _entries = new(StringComparer.Ordinal);
    private readonly object _sync = new();

    public IReadOnlyList<string> ActiveAppIds
    {
        get
        {
            lock (_sync) { return _entries.Keys.Order(StringComparer.Ordinal).ToArray(); }
        }
    }

    public PocketAppExecutionRuntime? Runtime(string appId)
    {
        lock (_sync)
        {
            return _entries.TryGetValue(appId, out var entry)
                ? entry.RuntimeHandle as PocketAppExecutionRuntime
                : null;
        }
    }

    public PocketAppRuntimeReadback? Readback(string appId)
    {
        lock (_sync) { return _entries.TryGetValue(appId, out var entry) ? entry.Readback : null; }
    }

    internal void Activate(
        PocketAppRuntimeReadback readback,
        object runtimeHandle,
        PocketAppActivationLease? activationLease)
    {
        lock (_sync)
        {
            if (_entries.Remove(readback.AppId, out var previous))
            {
                previous.ActivationLease?.Invalidate();
                DisposeRuntime(previous.RuntimeHandle);
            }
            _entries[readback.AppId] = new Entry(readback, runtimeHandle, activationLease);
        }
    }

    internal void Deactivate(string appId)
    {
        lock (_sync)
        {
            if (_entries.Remove(appId, out var previous))
            {
                previous.ActivationLease?.Invalidate();
                DisposeRuntime(previous.RuntimeHandle);
            }
        }
    }

    private static void DisposeRuntime(object runtimeHandle)
    {
        if (runtimeHandle is IDisposable disposable) { disposable.Dispose(); }
    }
}

internal sealed class PocketSurfaceRegistry
{
    internal sealed record Route(
        string AppId,
        string ProviderId,
        string SurfaceId,
        string Title);

    private sealed record Entry(
        PocketAppRuntimeReadback Readback,
        object HostHandle,
        IReadOnlySet<string> SurfaceIds);

    private readonly Dictionary<string, Entry> _entries = new(StringComparer.Ordinal);
    private readonly object _sync = new();

    public IReadOnlyList<string> ActiveAppIds
    {
        get
        {
            lock (_sync) { return _entries.Keys.Order(StringComparer.Ordinal).ToArray(); }
        }
    }

    public IReadOnlyList<Route> Routes
    {
        get
        {
            lock (_sync)
            {
                return _entries.Values
                    .Select(entry =>
                    {
                        var surfaceId = entry.SurfaceIds.Contains("main")
                            ? "main"
                            : entry.SurfaceIds.Order(StringComparer.Ordinal).FirstOrDefault();
                        if (surfaceId is null) { return null; }
                        var host = entry.HostHandle as PocketAppHostController;
                        return new Route(
                            entry.Readback.AppId,
                            GeneratedProviderId(entry.Readback.AppId),
                            surfaceId,
                            host?.AppName ?? entry.Readback.AppId);
                    })
                    .Where(route => route is not null)
                    .Cast<Route>()
                    .OrderBy(route => route.ProviderId, StringComparer.Ordinal)
                    .ToArray();
            }
        }
    }

    public PocketAppRuntimeReadback? Readback(string appId)
    {
        lock (_sync) { return _entries.TryGetValue(appId, out var entry) ? entry.Readback : null; }
    }

    public PocketAppHostController? HostController(string appId, string? surfaceId = null)
    {
        lock (_sync)
        {
            if (!_entries.TryGetValue(appId, out var entry)
                || (surfaceId is not null && !entry.SurfaceIds.Contains(surfaceId)))
            {
                return null;
            }
            return entry.HostHandle as PocketAppHostController;
        }
    }

    public static string GeneratedProviderId(string appId) =>
        $"generated-pocket-app:{appId}";

    public static string GeneratedSurfaceRouteId(string appId, string surfaceId) =>
        $"{GeneratedProviderId(appId)}/{surfaceId}";

    internal void Activate(
        PocketAppRuntimeReadback readback,
        object hostHandle,
        IReadOnlySet<string> surfaceIds)
    {
        lock (_sync) { _entries[readback.AppId] = new Entry(readback, hostHandle, surfaceIds); }
    }

    internal void Deactivate(string appId)
    {
        lock (_sync) { _entries.Remove(appId); }
    }
}

internal sealed class PocketAppRuntimeActivationRegistry : IDisposable
{
    internal sealed record Candidate(
        PocketAppRuntimeReadback Readback,
        object RuntimeHandle,
        object HostHandle,
        IReadOnlySet<string> SurfaceIds,
        PocketAppActivationLease? ActivationLease);

    private static readonly IReadOnlySet<string> ReservedAppIds =
        new HashSet<string>(["local.example.today-focus"], StringComparer.Ordinal);

    private readonly PocketAppLifecycleManager? _sourceLifecycle;
    private readonly Func<PocketAppManagementSnapshot> _managementSnapshotSource;
    private readonly Func<string, Candidate?> _candidateSource;
    private readonly Func<string, bool> _restoreFailurePersistence;
    private readonly Func<string, bool>? _failureInjection;
    private readonly object _activationSync = new();
    private bool _enabled = true;
    private bool _disposed;

    public PocketExecutionRuntimeRegistry ExecutionRegistry { get; } = new();
    public PocketSurfaceRegistry SurfaceRegistry { get; } = new();

    public bool TryGetManagedAppIds(out IReadOnlyList<string> appIds)
    {
        lock (_activationSync)
        {
            if (_disposed)
            {
                appIds = Array.Empty<string>();
                return false;
            }
            try
            {
                appIds = _managementSnapshotSource().Packages
                    .Where(package => package.State != PocketAppLifecycleState.Removed)
                    .Select(package => package.PackageId)
                    .Distinct(StringComparer.Ordinal)
                    .Order(StringComparer.Ordinal)
                    .ToArray();
                return true;
            }
            catch
            {
                appIds = Array.Empty<string>();
                return false;
            }
        }
    }

    public PocketAppRuntimeActivationRegistry(
        string rootDirectory,
        string userDataRoot,
        CapabilityBroker broker,
        string userId,
        Func<UserSettings> settings,
        Func<string, bool>? failureInjection = null)
    {
        var lifecycle = new PocketAppLifecycleManager(
            rootDirectory,
            userDataRoot,
            performStartupRecovery: true);
        _sourceLifecycle = lifecycle;
        _managementSnapshotSource = lifecycle.ManagementSnapshot;
        _candidateSource = packageId =>
        {
            var package = lifecycle.ActivePackageForActivation(packageId);
            if (package is null) { return null; }
            var effectivePermissions = package.Manifest.RequestedCapabilities
                .SelectMany(item => item.Permissions)
                .Distinct(StringComparer.Ordinal)
                .Order(StringComparer.Ordinal)
                .ToArray();
            PocketAppUserStateStore? stateStore = null;
            PocketAppActivationLease? activationLease = null;
            PocketAppExecutionRuntime? runtime = null;
            try
            {
                stateStore = new PocketAppUserStateStore(
                    package.Manifest.Id,
                    package.StatePropertyTypes,
                    userDataRoot);
                activationLease = new PocketAppActivationLease();
                runtime = new PocketAppExecutionRuntime(
                    package,
                    broker,
                    userId,
                    effectivePermissions.ToHashSet(StringComparer.Ordinal),
                    userStateStore: stateStore,
                    activationLease: activationLease);
                var host = new PocketAppHostController(runtime, settings);
                return new Candidate(
                    new PocketAppRuntimeReadback(
                        package.Manifest.Id,
                        package.Manifest.Version,
                        package.ManifestDigest,
                        effectivePermissions),
                    runtime,
                    host,
                    package.Surfaces.Keys.ToHashSet(StringComparer.Ordinal),
                    activationLease);
            }
            catch
            {
                activationLease?.Invalidate();
                if (runtime is not null) { runtime.Dispose(); }
                else { stateStore?.Dispose(); }
                throw;
            }
        };
        _restoreFailurePersistence = packageId =>
        {
            try
            {
                var receipt = lifecycle.Disable(packageId);
                var observed = lifecycle.DurableManagedPackage(packageId);
                return receipt.State == PocketAppLifecycleState.Disabled
                    && receipt.EffectivePermissions.Count == 0
                    && observed is not null
                    && observed.State == PocketAppLifecycleState.Disabled
                    && observed.Version == receipt.Version
                    && observed.PackageDigest == receipt.PackageDigest;
            }
            catch
            {
                return false;
            }
        };
        _failureInjection = failureInjection;
    }

    internal PocketAppRuntimeActivationRegistry(
        Func<IReadOnlyList<PocketAppManagedPackage>> managedPackagesSource,
        Func<string, Candidate?> candidateSource,
        Func<IReadOnlyList<PocketAppManagementIssue>>? managementIssuesSource = null,
        Func<string, bool>? restoreFailurePersistence = null,
        Func<string, bool>? failureInjection = null)
    {
        _managementSnapshotSource = () => new PocketAppManagementSnapshot(
            managedPackagesSource(),
            managementIssuesSource?.Invoke() ?? Array.Empty<PocketAppManagementIssue>());
        _candidateSource = candidateSource;
        _restoreFailurePersistence = restoreFailurePersistence ?? (_ => false);
        _failureInjection = failureInjection;
    }

    public PocketAppRuntimeReadback Synchronize(PocketAppLifecycleReceipt receipt)
    {
        lock (_activationSync)
        {
            return SynchronizeLocked(receipt);
        }
    }

    private PocketAppRuntimeReadback SynchronizeLocked(PocketAppLifecycleReceipt receipt)
    {
        ThrowIfDisposed();
        if (ReservedAppIds.Contains(receipt.PackageId))
        {
            FailClosed(receipt.PackageId);
            throw Failure("RUNTIME_ACTIVATION_RESERVED_ID");
        }

        if (receipt.State == PocketAppLifecycleState.Enabled)
        {
            if (!_enabled)
            {
                FailClosed(receipt.PackageId);
                throw Failure("RUNTIME_ACTIVATION_UNAVAILABLE");
            }
            var candidate = _candidateSource(receipt.PackageId);
            if (candidate is null)
            {
                FailClosed(receipt.PackageId);
                throw Failure("RUNTIME_ACTIVATION_READBACK_MISMATCH");
            }
            if (!candidate.Readback.Matches(receipt))
            {
                DisposeCandidate(candidate);
                FailClosed(receipt.PackageId);
                throw Failure("RUNTIME_ACTIVATION_READBACK_MISMATCH");
            }
            return Activate(candidate);
        }

        if (receipt.EffectivePermissions.Count != 0)
        {
            FailClosed(receipt.PackageId);
            throw Failure("RUNTIME_ACTIVATION_READBACK_MISMATCH");
        }
        FailClosed(receipt.PackageId);
        if (ExecutionRegistry.Readback(receipt.PackageId) is not null
            || SurfaceRegistry.Readback(receipt.PackageId) is not null)
        {
            throw Failure("RUNTIME_ACTIVATION_READBACK_MISMATCH");
        }
        return new PocketAppRuntimeReadback(
            receipt.PackageId,
            receipt.Version,
            receipt.PackageDigest,
            Array.Empty<string>());
    }

    public IReadOnlyList<string> RestoreEnabledApps()
    {
        lock (_activationSync)
        {
            return RestoreEnabledAppsLocked();
        }
    }

    private IReadOnlyList<string> RestoreEnabledAppsLocked()
    {
        ThrowIfDisposed();
        if (!_enabled) { return ["*"]; }
        PocketAppManagementSnapshot snapshot;
        try
        {
            snapshot = _managementSnapshotSource();
        }
        catch
        {
            return ["*"];
        }

        var failures = snapshot.Issues.Select(item => item.PackageId).ToList();
        foreach (var issue in snapshot.Issues)
        {
            FailClosed(issue.PackageId);
            _ = _restoreFailurePersistence(issue.PackageId);
        }
        foreach (var package in snapshot.Packages.OrderBy(item => item.PackageId, StringComparer.Ordinal))
        {
            if (package.State != PocketAppLifecycleState.Enabled)
            {
                FailClosed(package.PackageId);
                continue;
            }
            try
            {
                if (ReservedAppIds.Contains(package.PackageId))
                {
                    throw Failure("RUNTIME_ACTIVATION_RESERVED_ID");
                }
                var candidate = _candidateSource(package.PackageId);
                if (candidate is null
                    || candidate.Readback.AppId != package.PackageId
                    || candidate.Readback.Version != package.Version
                    || candidate.Readback.PackageDigest != package.PackageDigest)
                {
                    DisposeCandidate(candidate);
                    throw Failure("RUNTIME_ACTIVATION_READBACK_MISMATCH");
                }
                _ = Activate(candidate);
            }
            catch
            {
                FailClosed(package.PackageId);
                _ = _restoreFailurePersistence(package.PackageId);
                failures.Add(package.PackageId);
            }
        }
        return failures.Distinct(StringComparer.Ordinal).Order(StringComparer.Ordinal).ToArray();
    }

    public void Dispose()
    {
        lock (_activationSync)
        {
            if (_disposed) { return; }
            _enabled = false;
            RevokeAllLocked();
            _sourceLifecycle?.Dispose();
            _disposed = true;
        }
    }

    public void Shutdown()
    {
        SetEnabled(false);
    }

    public void SetEnabled(bool enabled)
    {
        lock (_activationSync)
        {
            ThrowIfDisposed();
            _enabled = enabled;
            if (!enabled)
            {
                RevokeAllLocked();
            }
        }
    }

    private void RevokeAllLocked()
    {
        foreach (var appId in ExecutionRegistry.ActiveAppIds
            .Concat(SurfaceRegistry.ActiveAppIds)
            .Distinct(StringComparer.Ordinal))
        {
            FailClosed(appId);
        }
    }

    private PocketAppRuntimeReadback Activate(Candidate candidate)
    {
        if (_failureInjection?.Invoke("before_runtime_registry_commit") == true)
        {
            FailClosed(candidate.Readback.AppId);
            DisposeCandidate(candidate);
            throw Failure("RUNTIME_ACTIVATION_UNAVAILABLE");
        }
        if (!_enabled)
        {
            FailClosed(candidate.Readback.AppId);
            DisposeCandidate(candidate);
            throw Failure("RUNTIME_ACTIVATION_UNAVAILABLE");
        }

        ExecutionRegistry.Activate(
            candidate.Readback,
            candidate.RuntimeHandle,
            candidate.ActivationLease);
        SurfaceRegistry.Activate(candidate.Readback, candidate.HostHandle, candidate.SurfaceIds);

        if (_failureInjection?.Invoke("runtime_readback_mismatch") == true
            || ExecutionRegistry.Readback(candidate.Readback.AppId) != candidate.Readback
            || SurfaceRegistry.Readback(candidate.Readback.AppId) != candidate.Readback)
        {
            FailClosed(candidate.Readback.AppId);
            throw Failure("RUNTIME_ACTIVATION_READBACK_MISMATCH");
        }
        return candidate.Readback;
    }

    private static void DisposeCandidate(Candidate? candidate)
    {
        if (candidate is null) { return; }
        candidate.ActivationLease?.Invalidate();
        if (candidate.RuntimeHandle is IDisposable disposable) { disposable.Dispose(); }
    }

    private void FailClosed(string appId)
    {
        SurfaceRegistry.Deactivate(appId);
        ExecutionRegistry.Deactivate(appId);
    }

    private void ThrowIfDisposed()
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
    }

    private static PocketAppRuntimeActivationException Failure(string code) => new(code);
}

File: Sources/HoverPocket/PocketApps/PocketAppRuntimeActivationRegistry.swift
import Combine
import Foundation

struct PocketAppRuntimeReadback: Equatable, Sendable {
    let appID: String
    let version: String?
    let packageDigest: String?
    let effectivePermissions: [String]

    init(
        appID: String,
        version: String?,
        packageDigest: String?,
        effectivePermissions: [String]
    ) {
        self.appID = appID
        self.version = version
        self.packageDigest = packageDigest
        self.effectivePermissions = effectivePermissions.sorted()
    }

    func matches(_ receipt: PocketAppLifecycleReceipt) -> Bool {
        appID == receipt.packageID
            && version == receipt.version
            && packageDigest == receipt.packageDigest
            && effectivePermissions == receipt.effectivePermissions.sorted()
    }
}

enum PocketAppRuntimeActivationError: Error, Equatable {
    case reservedIdentity
    case unavailable
    case readbackMismatch
}

@MainActor
final class PocketAppActivationLease {
    private(set) var isActive = true
    private var cancellationHandlers: [UUID: () -> Void] = [:]

    func requireActive() throws {
        guard isActive else { throw PocketAppRuntimeActivationError.unavailable }
    }

    func invalidate() {
        guard isActive else { return }
        isActive = false
        let handlers = Array(cancellationHandlers.values)
        cancellationHandlers.removeAll()
        handlers.forEach { $0() }
    }

    func registerCancellation(_ handler: @escaping () -> Void) -> UUID? {
        guard isActive else {
            handler()
            return nil
        }
        let id = UUID()
        cancellationHandlers[id] = handler
        return id
    }

    func unregisterCancellation(_ id: UUID?) {
        guard let id else { return }
        cancellationHandlers.removeValue(forKey: id)
    }
}

@MainActor
final class PocketExecutionRuntimeRegistry {
    private struct Entry {
        let readback: PocketAppRuntimeReadback
        let runtimeHandle: AnyObject
        let activationLease: PocketAppActivationLease?
    }

    private var entries: [String: Entry] = [:]

    var activeAppIDs: [String] { entries.keys.sorted() }

    func runtime(appID: String) -> PocketAppExecutionRuntime? {
        entries[appID]?.runtimeHandle as? PocketAppExecutionRuntime
    }

    func readback(appID: String) -> PocketAppRuntimeReadback? {
        entries[appID]?.readback
    }

    fileprivate func activate(
        _ readback: PocketAppRuntimeReadback,
        runtimeHandle: AnyObject,
        activationLease: PocketAppActivationLease?
    ) {
        entries[readback.appID]?.activationLease?.invalidate()
        entries[readback.appID] = Entry(
            readback: readback,
            runtimeHandle: runtimeHandle,
            activationLease: activationLease
        )
    }

    fileprivate func deactivate(appID: String) {
        entries.removeValue(forKey: appID)?.activationLease?.invalidate()
    }
}

@MainActor
final class PocketSurfaceRegistry: ObservableObject {
    struct Route: Equatable, Sendable {
        let appID: String
        let providerID: String
        let surfaceID: String
        let title: String
    }

    private final class Entry {
        private final class WeakModel {
            weak var value: PocketSurfaceHostModel?

            init(_ value: PocketSurfaceHostModel) {
                self.value = value
            }
        }

        let readback: PocketAppRuntimeReadback
        let runtimeHandle: AnyObject
        let surfaceIDs: Set<String>
        private var models: [String: [WeakModel]] = [:]

        init(
            readback: PocketAppRuntimeReadback,
            runtimeHandle: AnyObject,
            surfaceIDs: Set<String>
        ) {
            self.readback = readback
            self.runtimeHandle = runtimeHandle
            self.surfaceIDs = surfaceIDs
        }

        @MainActor
        func invalidate() {
            models.values
                .flatMap { $0 }
                .compactMap(\.value)
                .forEach { $0.invalidateActivation() }
            models.removeAll()
        }

        @MainActor
        func makeModel(surfaceID: String) throws -> PocketSurfaceHostModel? {
            guard let runtime = runtimeHandle as? PocketAppExecutionRuntime else { return nil }
            let model = try PocketSurfaceHostModel(runtime: runtime, surfaceID: surfaceID)
            models[surfaceID] = (models[surfaceID] ?? [])
                .filter { $0.value != nil } + [WeakModel(model)]
            return model
        }
    }

    private var entries: [String: Entry] = [:]
    @Published private(set) var revision = 0

    var activeAppIDs: [String] { entries.keys.sorted() }

    var routes: [Route] {
        entries.values.compactMap { entry in
            guard let surfaceID = entry.surfaceIDs.contains("main")
                    ? "main"
                    : entry.surfaceIDs.sorted().first else { return nil }
            let title = (entry.runtimeHandle as? PocketAppExecutionRuntime)?.package.manifest.name
                ?? entry.readback.appID
            return Route(
                appID: entry.readback.appID,
                providerID: Self.generatedProviderID(appID: entry.readback.appID),
                surfaceID: surfaceID,
                title: title
            )
        }.sorted { $0.providerID < $1.providerID }
    }

    func model(appID: String, surfaceID: String) throws -> PocketSurfaceHostModel? {
        guard let entry = entries[appID], entry.surfaceIDs.contains(surfaceID) else { return nil }
        return try entry.makeModel(surfaceID: surfaceID)
    }

    func readback(appID: String) -> PocketAppRuntimeReadback? {
        entries[appID]?.readback
    }

    nonisolated static func generatedProviderID(appID: String) -> String {
        "generated-pocket-app:\(appID)"
    }

    nonisolated static func generatedAppID(providerID: String) -> String? {
        let prefix = "generated-pocket-app:"
        guard providerID.hasPrefix(prefix) else { return nil }
        let appID = String(providerID.dropFirst(prefix.count))
        guard appID.unicodeScalars.count <= 160,
              appID.range(
                  of: "^[a-z][a-z0-9]*(?:\\.[a-z0-9][a-z0-9-]*){2,}$",
                  options: .regularExpression
              ) != nil else {
            return nil
        }
        return appID
    }

    nonisolated static func generatedSurfaceRouteID(appID: String, surfaceID: String) -> String {
        "\(generatedProviderID(appID: appID))/\(surfaceID)"
    }

    func activate(
        _ readback: PocketAppRuntimeReadback,
        runtimeHandle: AnyObject,
        surfaceIDs: Set<String>
    ) {
        entries.removeValue(forKey: readback.appID)?.invalidate()
        entries[readback.appID] = Entry(
            readback: readback,
            runtimeHandle: runtimeHandle,
            surfaceIDs: surfaceIDs
        )
        revision &+= 1
    }

    func deactivate(appID: String) {
        guard let entry = entries.removeValue(forKey: appID) else { return }
        entry.invalidate()
        revision &+= 1
    }
}

@MainActor
final class PocketAppRuntimeActivationRegistry {
    struct Candidate {
        let readback: PocketAppRuntimeReadback
        let runtimeHandle: AnyObject
        let activationLease: PocketAppActivationLease?
        let surfaceIDs: Set<String>
    }

    typealias ManagementSnapshotSource = () throws -> PocketAppManagementSnapshot
    typealias ManagedPackagesSource = () throws -> [PocketAppManagedPackage]
    typealias ManagementIssuesSource = () throws -> [PocketAppManagementIssue]
    typealias CandidateSource = (String) throws -> Candidate?
    typealias RestoreFailurePersistence = (String) -> Bool

    let executionRegistry = PocketExecutionRuntimeRegistry()
    let surfaceRegistry = PocketSurfaceRegistry()

    private let sourceLifecycle: PocketAppLifecycleManager?
    private let managementSnapshotSource: ManagementSnapshotSource
    private let candidateSource: CandidateSource
    private let restoreFailurePersistence: RestoreFailurePersistence
    private let failureInjection: ((String) -> Bool)?
    private let reservedAppIDs: Set<String>

    init(
        rootDirectory: URL,
        userDataRoot: URL,
        broker: CapabilityBroker,
        userID: String,
        failureInjection: ((String) -> Bool)? = nil
    ) throws {
        let lifecycle = try PocketAppLifecycleManager(
            rootDirectory: rootDirectory,
            userDataRoot: userDataRoot,
            performStartupRecovery: true
        )
        self.sourceLifecycle = lifecycle
        self.managementSnapshotSource = lifecycle.managementSnapshot
        self.candidateSource = { packageID in
            guard let package = try lifecycle.activePackageForActivation(packageID: packageID) else {
                return nil
            }
            let effectivePermissions = package.manifest.requestedCapabilities.reduce(into: Set<String>()) {
                $0.formUnion($1.permissions)
            }
            let stateStore = try PocketAppUserStateStore(
                packageID: package.manifest.id,
                propertyTypes: package.statePropertyTypes,
                rootDirectory: userDataRoot
            )
            let activationLease = PocketAppActivationLease()
            let runtime = PocketAppExecutionRuntime(
                package: package,
                broker: broker,
                userID: userID,
                grantedPermissions: effectivePermissions,
                userStateStore: stateStore,
                activationLease: activationLease
            )
            return Candidate(
                readback: PocketAppRuntimeReadback(
                    appID: package.manifest.id,
                    version: package.manifest.version,
                    packageDigest: package.manifestDigest,
                    effectivePermissions: effectivePermissions.sorted()
                ),
                runtimeHandle: runtime,
                activationLease: activationLease,
                surfaceIDs: Set(package.surfaces.keys)
            )
        }
        self.restoreFailurePersistence = { packageID in
            do {
                let receipt = try lifecycle.disable(packageID: packageID)
                guard let observed = try lifecycle.durableManagedPackage(packageID: packageID) else {
                    return false
                }
                return receipt.state == .disabled
                    && receipt.effectivePermissions.isEmpty
                    && observed.state == .disabled
                    && observed.version == receipt.version
                    && observed.packageDigest == receipt.packageDigest
            } catch {
                return false
            }
        }
        self.failureInjection = failureInjection
        self.reservedAppIDs = ["local.example.today-focus"]
    }

    init(
        managedPackagesSource: @escaping ManagedPackagesSource,
        managementIssuesSource: @escaping ManagementIssuesSource = { [] },
        candidateSource: @escaping CandidateSource,
        reservedAppIDs: Set<String> = ["local.example.today-focus"],
        restoreFailurePersistence: @escaping RestoreFailurePersistence = { _ in false },
        failureInjection: ((String) -> Bool)? = nil
    ) {
        self.sourceLifecycle = nil
        self.managementSnapshotSource = {
            PocketAppManagementSnapshot(
                packages: try managedPackagesSource(),
                issues: try managementIssuesSource()
            )
        }
        self.candidateSource = candidateSource
        self.restoreFailurePersistence = restoreFailurePersistence
        self.failureInjection = failureInjection
        self.reservedAppIDs = reservedAppIDs
    }

    @discardableResult
    func synchronize(_ receipt: PocketAppLifecycleReceipt) throws -> PocketAppRuntimeReadback {
        guard !reservedAppIDs.contains(receipt.packageID) else {
            failClosed(appID: receipt.packageID)
            throw PocketAppRuntimeActivationError.reservedIdentity
        }

        switch receipt.state {
        case .enabled:
            guard let candidate = try candidateSource(receipt.packageID),
                  candidate.readback.matches(receipt) else {
                failClosed(appID: receipt.packageID)
                throw PocketAppRuntimeActivationError.readbackMismatch
            }
            return try activate(candidate, expected: candidate.readback)

        case .disabled, .removed:
            guard receipt.effectivePermissions.isEmpty else {
                failClosed(appID: receipt.packageID)
                throw PocketAppRuntimeActivationError.readbackMismatch
            }
            failClosed(appID: receipt.packageID)
            guard executionRegistry.readback(appID: receipt.packageID) == nil,
                  surfaceRegistry.readback(appID: receipt.packageID) == nil else {
                throw PocketAppRuntimeActivationError.readbackMismatch
            }
            return PocketAppRuntimeReadback(
                appID: receipt.packageID,
                version: receipt.version,
                packageDigest: receipt.packageDigest,
                effectivePermissions: []
            )
        }
    }

    @discardableResult
    func restoreEnabledApps() -> [String] {
        let snapshot: PocketAppManagementSnapshot
        do {
            snapshot = try managementSnapshotSource()
        } catch {
            return ["*"]
        }

        var failures = snapshot.issues.map(\.packageID)
        for issue in snapshot.issues {
            failClosed(appID: issue.packageID)
            _ = restoreFailurePersistence(issue.packageID)
        }
        for package in snapshot.packages.sorted(by: { $0.packageID < $1.packageID }) {
            guard package.state == .enabled else {
                failClosed(appID: package.packageID)
                continue
            }
            do {
                guard !reservedAppIDs.contains(package.packageID),
                      let candidate = try candidateSource(package.packageID),
                      candidate.readback.appID == package.packageID,
                      candidate.readback.version == package.version,
                      candidate.readback.packageDigest == package.packageDigest else {
                    throw PocketAppRuntimeActivationError.readbackMismatch
                }
                _ = try activate(candidate, expected: candidate.readback)
            } catch {
                failClosed(appID: package.packageID)
                _ = restoreFailurePersistence(package.packageID)
                failures.append(package.packageID)
            }
        }
        return Array(Set(failures)).sorted()
    }

    func managedAppIDs() throws -> Set<String> {
        Set(
            try managementSnapshotSource().packages.compactMap { package in
                package.state == .removed ? nil : package.packageID
            }
        )
    }

    func shutdown() {
        let appIDs = Set(executionRegistry.activeAppIDs).union(surfaceRegistry.activeAppIDs)
        for appID in appIDs {
            failClosed(appID: appID)
        }
    }

    private func activate(
        _ candidate: Candidate,
        expected: PocketAppRuntimeReadback
    ) throws -> PocketAppRuntimeReadback {
        if failureInjection?("before_runtime_registry_commit") == true {
            failClosed(appID: expected.appID)
            throw PocketAppRuntimeActivationError.unavailable
        }

        executionRegistry.activate(
            expected,
            runtimeHandle: candidate.runtimeHandle,
            activationLease: candidate.activationLease
        )
        surfaceRegistry.activate(
            expected,
            runtimeHandle: candidate.runtimeHandle,
            surfaceIDs: candidate.surfaceIDs
        )

        if failureInjection?("runtime_readback_mismatch") == true
            || executionRegistry.readback(appID: expected.appID) != expected
            || surfaceRegistry.readback(appID: expected.appID) != expected {
            failClosed(appID: expected.appID)
            throw PocketAppRuntimeActivationError.readbackMismatch
        }

        return expected
    }

    private func failClosed(appID: String) {
        surfaceRegistry.deactivate(appID: appID)
        executionRegistry.deactivate(appID: appID)
    }
}
